Added a ScoreSystem to handle player scores.

zephyyrr opened the pull request:
New ScoreSystem class. Implements the various scoring events and deals out points based on performance.

Score is rendered in the UI next to the current level.

Exact scoring algorithm for final game is still under consideration.
If further scoring events is required, add as necessary.

zephyyrr commented:
So in order to make this much better, we need to implement #28 first.

Victorystick commented on src/ScoreSystem.js:
```diff
@@ -0,0 +1,27 @@
+
+export class ScoreSystem {
+	constructor() {
+  }
```
Indentation!

Victorystick commented on src/ScoreSystem.js:
```diff
@@ -0,0 +1,27 @@
+
+export class ScoreSystem {
+	constructor() {
```
You can omit the constructor if
* it is empty
* it delegates all arguments to the super-constructor

Victorystick replied:
Although why not just create a score keeping counter for a specific player character? If we move the player character creation to Logic, it can create and manage the scorekeeping stuff there.

Victorystick commented on src/ScoreSystem.js:
```diff
@@ -0,0 +1,27 @@
+  }
+
+	initialize(player) {
+		player.score = 0;
```
Although tempting because of how easy it is, you shouldn't add properties to an object after its construction. It prevents the JIT from performing certain optimisations.

Victorystick commented on src/ScoreSystem.js:
```diff
@@ -0,0 +1,27 @@
+	}
+
+	updateScore(player, dscore) {
+		if (player != null) {
```
You can make this even easier. `if (player) { ...` will do.

zephyyrr commented:
Yeah. Because a uninitialized player's score is undefined. So I just reverted to the original Map's implementation without the null check I added when I started overthinking the problem.
